Assets/MRTK/Services/CameraSystem/ProjectionOverride.cs

```diff
@@ -0,0 +1,114 @@
+    /// So there is no need for an AddComponentMenu attribute.
+    /// </remarks>
+    internal class ProjectionOverride : MonoBehaviour
+    {
+        /// <summary>
+        /// When this is true, projection will be overridden on each frame
+        /// </summary>
+        public bool EnableOverride { get; set; } = false;
+
+#if WINDOWS_UWP
+        private struct InternalHoloLensFrameStructure
+        {
+#pragma warning disable 0649
+            public uint VersionNumber;
+            public uint MaxNumberOfCameras;
+            public IntPtr ISpatialCoordinateSystemPtr;
+            public IntPtr IHolographicFramePtr;
+            public IntPtr IHolographicCameraPtr;
+#pragma warning restore 0649
+        }
+#endif
+
+        /// <summary>
+        /// Coroutine function to set the camera matrices back to their defaults
+        /// </summary>
+        /// <returns>Coroutine enumerator</returns>
+        IEnumerator ResetViewMatricesOnFrameEnd()
+        {
+            yield return new WaitForEndOfFrame();
+            Camera cam = GetComponent<Camera>();
+            cam.ResetStereoViewMatrices();
+            cam.ResetStereoProjectionMatrices();
+        }
+
+        /// <inheritdoc />
+        void OnPreCull()
+        {
+            if (!EnableOverride)
+                return;
+
+            StartCoroutine(ResetViewMatricesOnFrameEnd());
+
+#if WINDOWS_UWP
+            IntPtr nativeStruct = UnityEngine.XR.XRDevice.GetNativePtr();
+            if (nativeStruct != IntPtr.Zero)
+            {
+                InternalHoloLensFrameStructure s = System.Runtime.InteropServices.Marshal.PtrToStructure<InternalHoloLensFrameStructure>(nativeStruct);
+
+                if (s.IHolographicFramePtr != IntPtr.Zero)
+                {
+                    var holographicCamera = (HolographicCamera)System.Runtime.InteropServices.Marshal.GetObjectForIUnknown(s.IHolographicCameraPtr);
+                    var holographicFrame = (HolographicFrame)System.Runtime.InteropServices.Marshal.GetObjectForIUnknown(s.IHolographicFramePtr);
+                    HolographicFramePrediction prediction = holographicFrame.CurrentPrediction;
+                    for (int i = 0; i < prediction.CameraPoses.Count; ++i)
+                    {
+                        HolographicCameraPose cameraPose = prediction.CameraPoses[i];
+
+                        // Default spatial locator is the HMD
+                        SpatialLocator defaultSpatialLocator = SpatialLocator.GetDefault();
+
+                        SpatialLocatorAttachedFrameOfReference attachedFrameOfReference = defaultSpatialLocator.CreateAttachedFrameOfReferenceAtCurrentHeading();
+                        SpatialCoordinateSystem attachedCoordinateFrame = attachedFrameOfReference.GetStationaryCoordinateSystemAtTimestamp(prediction.Timestamp);
+
+                        HolographicStereoTransform stereoProjection = cameraPose.ProjectionTransform;
+                        if (holographicCamera.CanOverrideViewport)
```

Comment:
It feels like this check should happen before line 74 right? Otherwise we'll do a bunch of work each time and end up no-oping deep within each for loop.

Or is it the case that holographicCamera is mutated through some of the Camera.main or cameraPose calls?